RFC: Enhanced Ansible support (#98)
uyuni-project/uyuni-rfc: accepted/00104-enhanced-ansible-support.md

We identified the following improvements to enhance the current implementation:

1) Show the raw output results of a playbook in the event history of the control node along with a list of targeted inventory systems.
2) Add UI support to allow editing of variables defined in the playbook.
3) Support the recurrent scheduling of Ansible playbooks.
4) Set default playbook and inventory paths when adding a new control node in Uyuni.
5) Add option to filter by Ansible managed systems onboarded in Uyuni in the system list.

These points will be discussed one-by-one in detail in the next section.

# Detailed design
[design]: #detailed-design

## Show the raw output results of a playbook in the event history of the control node along with a list of targeted inventory systems.

#### Raw output

To show the raw output returned by the ansible command itself instead of the output returned by salt a change to the [runplaybook.sls](https://github.com/uyuni-project/uyuni/blob/master/susemanager-utils/susemanager-sls/salt/ansible/runplaybook.sls) will have to be made.

We're currently using the `state module`:
```
run_ansible_playbook:
  ansible.playbooks:
     - .......
```

> To get the raw output we will have to use module.run to call the `ansible.playbooks` function from the `execution module`:

Reviewer: I would expect this would't have any performance downsides.

```
run_ansible_playbook:
  module.run:
    - name: ansible.playbooks
    - ....
```

The changed `runplaybook.sls` would then look like:
```
run_ansible_playbook:
  mgrcompat.module_run:
    - name: ansible.playbooks
    - playbook: {{ pillar["playbook_path"] }}
    - rundir: {{ pillar["rundir"] }}
    - flush_cache: {{ pillar["flush_cache"] }}
{%- if "inventory_path" in pillar %}
    - inventory: {{ pillar["inventory_path"] }}
{% endif %}
```

#### Displaying targeted systems

Since the list of systems targeted by an ansible playbook can be very long, I propose to display the used inventory as a link in the event history details instead.

In case the selected event details are of the playbook action type we will set an additional request parameter called `inventory` in `SystemHistoryEventAction.java` and render an extra row that would look like:
```
Inventory: <link_to_inventory>
```

> Clicking the link would then forward the user to the inventory file in `System` > `Ansible` > `Inventories`.

Reviewer: It's a tricky one. So far, we haven't been keeping track of any systems of inventory. Yes, currently, if inventory is removed from the control node, we will have an orphan inventory link in our database IIRC.

I would suggest to just make this targeted systems part of playbook action by parsing the output or inventory at schedule time. Keeping it linked with inventory will create the issue what you just brought up, Ricardo. Yes, it's understood that list could be long.

> Clicking the link would then forward the user to the inventory file in `System` > `Ansible` > `Inventories`.

Reviewer 2: Abid I think that is a good option.
A possible alternative is to save the inventory link/name at schedule time, and then at execution time we save the file content in the database, for example as part of the execution result. 
I'm not sure what would be better, but one way or the other we should save the inventory list without link to the file.

The main good point of just saving in the result is for recurrent executions if users change the inventory file it will reflect be reflected in the next recurrent execution, and as user I would expect that to happen.

> Clicking the link would then forward the user to the inventory file in `System` > `Ansible` > `Inventories`.

Reviewer 2: I like option 2 more, it's more clean and direct.

The URL to load a specific inventory would look like:
```
manager/systems/details/ansible/inventories?sid=<system_id>&iid=<inventory_id>
```
with `<system_id>` being the system id of the ansible control node and
`<inventory_id` being the id of the inventory to load.

## Add UI support to allow editing of variables defined in the playbook

The plan here is to enhance the existing UI in `System > Ansible > Playbooks > select playbook` with generated input fields based on the `vars` defined in the playbook.

For example
```
- name: Example playbook with various variable types
  hosts: all
  vars:
    # List variable
    software_packages:
      - git
      - nginx
      - curl

    # Dictionary variable
    user_info:
      username: "devuser"
      uid: 1001
      home_dir: "/home/devuser"

    # Integer variable
    max_open_files: 1024

    # Boolean variable
    enable_firewall: true

    # String variable
    welcome_message: "Welcome to the server setup!"
```

> The UI would generate input fields based on variable type (list, dict, string) that are pre-populated with the default variables set in the playbook (if any).

Reviewer 3: Since we already have a fair bit of code that does many similar things for formulas, do you think we could extend/reuse that, or do you think the underlying requirements are too different?

Reviewer: Underlaying requirements aren't that different. If we could reuse some exiting components that would be ideal. My only concern is that formulas with forms code was pretty fragile and painful to maintain in the past, not sure if we want to build more stuff on top of it.

Reviewer 2: The front-end I think was not that bad, only the hide/show part was not obvious (but I think Karl/Richa have work or at least plans to improve it). The main problem is the formulas overall integration

Changing the values of the generated input fields will allow users to override the defaults and schedule the playbook using the variables set in said UI.

> Additionally we should add a `free-form` text field to allow users to define additonal variables that were not part of the `vars` section and thus don't have an input field generated. This input would have to be in JSON format.

Reviewer: Instead of free-form, would it be feasible to have some  UI where user can add the variables dynamically , we already know the possible types of the variables . So for user, it will be choosing the right type and then add the values as suitable/needed?

On a general note, in Ansible, YAML is far more commonly used than JSON so my preference would be always to have something that users are used if needed.

> Additionally we should add a `free-form` text field to allow users to define additonal variables that were not part of the `vars` section and thus don't have an input field generated. This input would have to be in JSON format.

Reviewer 2: @parlt91 yaml extra extra_args (the ones from free text) only need to be parsed and merged when passing them to ansbile for execution. You don't want to merge then on the front-end and pass all together to the backend, otherwise, you lose track of what is on the fields, and what is on the free form.

> Based on the variables we will generate a JSON string called `extra_vars` that will be send to the backend along with other parameters needed to schedule the playbook.

Reviewer: Agree! This goes in the same direction as systems of inventory so I would suggest fixing this problem together. When a playbook executed, we need to know exactly what variables were passed and which systems were targeted - this information should be possible to get at any time after the execution is done!

The workflow on the frontend would look like:
1) Parse the selected playbook.
2) Generate input fields based on the `vars` section of the playbook.
3) Compile a JSON string from the inputs that will be send to the backend.

> The `ansible.playbooks` salt state we use to execute playbooks at the scheduled time supports a parameter called `extra_vars` and accepts a JSON string. Variables set through this parameter will take precedence over the ones defined in the playbook and thus allow us to override them.

Reviewer 4: That's right, but then IIUC we should rephrase this to not really mention "salt state" here in this sentence. Maybe something like:

```suggestion
The `ansible.playbooks` salt execution module function we use to execute playbooks at the scheduled time supports a parameter called `extra_vars` and accepts a JSON string. Variables set through this parameter will take precedence over the ones defined in the playbook and thus allow us to override them.
```

> 

Reviewer: Ok so we will enable user to set the vars of the palybook and then if that user also set the same variable additionally, by mistake or so, under as part of our freeform idea,  which one will take precedence? or will we add some validation that if variable is already defined, it shouldn't be possible to define it again?

Reviewer 2: I would add this extra variable in json/yaml/free text mode under an advance configuration option. Which one would have precedence I don't have a strong opinion. Maybe the advanced ones? over the fields?
Or maybe the input field one, which should make it easier to identify where the value is coming from.

To support scheduling playbooks using `extra_vars` on the backend we will have to to add a new column to the `rhnActionPlaybook` table and adapt the corresponding hibernate entity. The JSON string could either be stored as `varchar` or alternatively as `blob` if we expect them to be large.

> During execution of the action the `extra_vars` will then be passed as pillar data to the `runplaybook.sls` state that then executes the `ansible.playbooks` state using the pillar data.

Reviewer 4: ```suggestion
During execution of the action the `extra_vars` will then be passed as pillar data to the `runplaybook.sls` state that then executes the `ansible.playbooks` execution module function using the pillar data.
```

    - inventory: {{ pillar["inventory_path"] }}
{% endif %}
```

#### Storing of variables edited by the user

To avoid that users have to fill the input field with their custom changes every time, we could implement a mechanism that would store their last used configuration.

To support this I propose adding a new table called e.g. `suseAnsiblePlaybookVars` with the following columns:

- `server_id`
- `user_id`
- `playbook_path`
- `extra_vars`

with the unique identifier being the combination of `server_id`, `user_id` and `playbook_path`.

This data would be stored whenever a playbook is scheduled from the UI.
If there is already a stored config with matching `server_id`, `user_id` and `playbook_path` we would override the previous one.

When loading the playbook scheduling page we would look up, if there is an existing config and merge the data into the generated input fields. Fields without a match (likely because the playbook has been edited by the user) would be omitted.

Along with this we should also provide a reset button to restore the input fields to the playbook defined defaults.

## Support the recurrent scheduling of Ansible playbooks.

> To support the execution of Ansible playbooks using recurring actions we'll have to implement the follwoing steps:

Reviewer 4: ```suggestion
To support the execution of Ansible playbooks using recurring actions we'll have to implement the following steps:
```

To get the list of systems managed by an Ansible control node we'll scan all the inventories registered in `suseAnsiblePath`.

There is currently a salt state called `ansible.targets inventory=<some_inventory>` available that returns inventory data. The list of systems can be parsed from that data (this mechanism was already part of the original Ansible implementation).

However there are two problems that need two be solved here:
1) We need a way to automate collecting Ansible managed systems to make sure the list is up to date.

    This can be achieved by adding a `inotify beacon` on every control node. This beacon would contain the list of inventories registered on Uyuni and trigger an event to update the list of systems whenever one of the inventory files receives an update.

    See example beacon below:
    ```
    beacons:
      inotify:
        - files:
            /etc/ansible/hosts:
              mask:
                - modify
            /another/inventory/file:
              mask:
               - modify
        - interval: 5
        - disable_during_state_run: True
    ```

    This beacon would be updated whenever a inventory file is added/removed on Uyuni.

>     Along with the beacon we'll have to add a new `reactor` to the salt master that will react to events send by the beacons.

Reviewer 2: What would this reactor do? Shouldn't we just process the event? Save it on the database and let the salt event processor handle it?

Reviewer 4: Yeah, I think it is better to use the `SaltReactor.java` here and not the Salt Master Reactors.

              hosts:
                  - some-minion.tf.local
          _meta:
              ----------
              hostvars:
                  ----------
          all:
              ----------
              children:
                  - Some_group
                  - ungrouped
      <another_inventory>:
          ----------
          Another_group:
              ----------
              hosts:
                  - another-minion.tf.local
                  - additional hosts
          _meta:
              ----------
              hostvars:
                  ----------
          all:
              ----------
              children:
                  - Another_group
                  - ungrouped
      <yet_another_one>:
        .
        .
        .
    ```
    This way we'd only have to do one single salt execution per control node. We would then parse the output and compile it into a list of systems. This list will be used to add/remove the `ansible_managed` entitlement from systems registered in Uyuni.

# Drawbacks
[drawbacks]: #drawbacks

I don't currently see any drawback with the proposed implementation.

# Future improvements
- Make Ansible work with Uyunis system groups.

> - Allow adding Ansible playbooks to custom state recurring actions to support executing multiple playbooks in order.

Reviewer 2: Or have this feature directly on ansible playbook, and not integrated into the custom states. However, we need to be careful to not re-implement awx.

# Unresolved questions
[unresolved]: #unresolved-questions

The following points still need to be investigated in:

> - It looks like the `inotify beacon` requires `python3-inotify package` to be installed on every control node. Is this beacon already available by other means or do we have to make sure the package is installed when setting up a new control node in Uyuni?

Reviewer 4: When we set a system as "Ansible Control Node" (in system properties), then this injects some bits to the system highstate, particularly it forces `ansible` package to be installed, so we can also make some other packages to be installed.

But an important thing here is that, beacons would be executed by `venv-salt-minion` (as in 5.0+ all registered clients are migrated to "venv-salt-minion"), so we need python3-inotify for the Salt Bundle. The good thing is that the Salt Bundle already contains `saltbundlepy-inotify`, meaning this will be available in all Salt Bundle supported systems.

Another thing worth to mention in this RFC that beacons won't work in Salt SSH, so this approach won't work for Ansible Control Nodes managed via Salt SSH.

> - How to update the `inotify` beacon from the webUI when new inventories added/removed?

Reviewer 4: The beacon definition in Uyuni is done via pillar data for the minions. So Uyuni provides this "beacon" configuration as part of the generated pillar data for a particular minion.

My understanding here is that, whenever you add/remove an inventory from the UI, this should update the generated pillar data for this minion to include the new  beacon configuration and then trigger 
 a `saltutil.refresh_pillar` and `saltutil.refresh_beacons` actions for the minion, so it refreshes the beacons on the running salt minion process.

> - How will the results of the `ansible.targets` state triggered through above beacon be handled by the java backend?

Reviewer 4: As clarified in previous comments, this would be done by `SaltReactor.java`.

To get the list of systems managed by an Ansible control node we'll scan all the inventories registered in `suseAnsiblePath`.

There is currently a salt state called `ansible.targets inventory=<some_inventory>` available that returns inventory data. The list of systems can be parsed from that data (this mechanism was already part of the original Ansible implementation).

However there are two problems that need two be solved here:
1) We need a way to automate collecting Ansible managed systems to make sure the list is up to date.

    This can be achieved by adding a `inotify beacon` on every control node. This beacon would contain the list of inventories registered on Uyuni and trigger an event to update the list of systems whenever one of the inventory files receives an update.

    See example beacon below:
    ```
    beacons:
      inotify:
        - files:
            /etc/ansible/hosts:
              mask:
                - modify
            /another/inventory/file:
              mask:
               - modify
        - interval: 5
        - disable_during_state_run: True
    ```

>     The beacons will be executed by `venv-salt-minion`. To add support for the `inotify` beacon to them we'll have to make sure the `saltbundlepy-inotify` package is installed along the `ansible` package whenever a new control node is added. Already existing control nodes will have to run a highstate to get the package installed.

Reviewer 4: No, already in 5.0, the registered clients are automatically migrated to `venv-salt-minion`, if not using it yet, as part of the highstate.

    The beacon would be updated whenever a inventory file is added/removed on Uyuni. To support this we will add the beacon to the minions pillar data and use the `MinionPillarManager.generatePillar(...)` function to update it whenever needed. In addition we will have to run `saltutil.refresh_pillar` and `saltutil.refresh_beacons` to refresh data on the running minion process

    On the java backend we will add support for the beacon to the `SaltReactor.java` and trigger an update of the ansible managed systems whenever there were changes to the inventory files.

    Additionally we'll trigger a manual update of the minion list whenever inventories are added/removed.

>     One thing to mention is that this approach (using the beacon) will `NOT` work with ansible control notes managed through Salt SSH.

Reviewer 4: Small typo:
```suggestion
    One thing to mention is that this approach (using the beacon) will `NOT` work with ansible control nodes managed through Salt SSH.
```
